tests/Microsoft.Identity.Web.Test/WebAppExtensionsTests.cs

```diff
@@ -1,32 +1,493 @@
+        {
+            var configMock = Substitute.For<IConfiguration>();
+            configMock.Configure().GetSection(_configSectionName).Returns(_configSection);
+
+            var diagnosticsMock = Substitute.For<IOpenIdConnectMiddlewareDiagnostics>();
+
+            var services = new ServiceCollection();
+            services.AddDataProtection();
+
+            if (useServiceCollectionExtension)
+                services.AddSignIn(configMock, _configSectionName, _oidcScheme, _cookieScheme, subscribeToDiagnostics);
+            else
+                new AuthenticationBuilder(services)
+                    .AddSignIn(configMock, _configSectionName, _oidcScheme, _cookieScheme, subscribeToDiagnostics);
+
+            AddSignIn_TestSubscribesToDiagnostics(services, diagnosticsMock, subscribeToDiagnostics);
+        }
+
+        [Theory]
+        [InlineData(true, true)]
+        [InlineData(true, false)]
+        [InlineData(false, true)]
+        [InlineData(false, false)]
+        public void AddSignIn_WithConfigActions_SubscribesToDiagnostics(bool useServiceCollectionExtension, bool subscribeToDiagnostics)
+        {
+            var diagnosticsMock = Substitute.For<IOpenIdConnectMiddlewareDiagnostics>();
+
+            var services = new ServiceCollection();
+            services.AddDataProtection();
+
+            if (useServiceCollectionExtension)
+                services.AddSignIn(_configureOidcOptions, _configureMsOptions, _oidcScheme, _cookieScheme, subscribeToDiagnostics);
+            else
+                new AuthenticationBuilder(services)
+                    .AddSignIn(_configureOidcOptions, _configureMsOptions, _oidcScheme, _cookieScheme, subscribeToDiagnostics);
+
+            AddSignIn_TestSubscribesToDiagnostics(services, diagnosticsMock, subscribeToDiagnostics);
+        }
+
+        private void AddSignIn_TestSubscribesToDiagnostics(IServiceCollection services, IOpenIdConnectMiddlewareDiagnostics diagnosticsMock, bool subscribeToDiagnostics)
+        {
+            services.RemoveAll<IOpenIdConnectMiddlewareDiagnostics>();
+            services.AddSingleton<IOpenIdConnectMiddlewareDiagnostics>((provider) => diagnosticsMock);
+
+            var provider = services.BuildServiceProvider();
+
+            var oidcOptions = provider.GetRequiredService<IOptionsFactory<OpenIdConnectOptions>>().Create(_oidcScheme);
+
+            // Assert subscribed to diagnostics
+            if (subscribeToDiagnostics)
+                diagnosticsMock.ReceivedWithAnyArgs().Subscribe(Arg.Any<OpenIdConnectEvents>());
+            else
+                diagnosticsMock.DidNotReceiveWithAnyArgs().Subscribe(Arg.Any<OpenIdConnectEvents>());
+        }
+
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public async Task AddSignIn_WithConfigName_B2cSpecificSetup(bool useServiceCollectionExtension)
+        {
+            var configMock = Substitute.For<IConfiguration>();
```

Comment:
there shouldn't be too many differences between b2c and AAD set up. could you have an additional bool, like `isB2C` and then route accordingly? This is basically the same as AddSIgnIn_WithConfigName()

Reply:
Well unit tests should test one thing, have no logic, be simple, act as documentation. This way if this one unit test breaks but the rest are fine, I know something related to B2C broke. Writing tests this way does create a ton of duplicated quote (partially because setup code can't be abstracted away because of small differences). On the other extreme I can write just one 250+ line method to test everything. This reduces the code duplication but adds a potential for errors and, if it breaks, I have to debug it, basically.

I haven't strictly followed the [best practices](AddSignIn_TestOnRedirectToIdentityProviderEvent) but tried to strike a balance between the two extremes whenever it made sense. You can look at [AadIsserValidatorTests](https://github.com/AzureAD/microsoft-identity-web/blob/master/tests/Microsoft.Identity.Web.Test/Resource/AadIssuerValidatorTests.cs) where I took the opposite approach of writing smaller tests with more duplication.

So in a sense, since both extremes are not perfect, I'm okay with writing tests that more or less makes sense and work and we can refactor them later, if needed. (The 80-20 rule).

@jmprieur @henrik-me thoughts?

Reply:
I get that for the individual test, I guess I mean more for the private methods that seem to have very little difference between them. Or just move the configuration that's the same in these public methods somewhere else to remove "noise". just a suggestion.

---
In reply to: [417813473](https://github.com/AzureAD/microsoft-identity-web/pull/132#discussion_r417813473) [](ancestors = 417813473)

Reply:
I"ll let @bgavrilMS provide his expertise (#not-a-dev)